### PR TITLE
#4732: refresh partner auth token in the background

### LINKED ISSUE #4732: Use refresh token from partner OAuth

### PR DESCRIPTION
**Pull request by twschiller**
## What does this PR do?

- Closes #4732 
- If using an AA JWT with a refresh_token, automatically refresh the token every 10 min

## Discussion

- On refresh failure (e.g., due to expiration), the code doesn't clear out the token state
- If the JWT is expired, the next time the JWT is used to authenticate against our API or the control room, it will be cleared out, prompting re-authentication

## Demo

- TBD

## TODO

- [ ] manually perform E2E test and create demo of network inspector

## Checklist

- [x] Add tests
- [x] Run Storybook and manually confirm that all stories are working: N/A
- [x] Designate a primary reviewer: @mnholtz

### CODE REVIEW

File: src/background/background.ts [hunk 1]
```diff
@@ -59,3 +60,4 @@ activateBrowserActionIcon();
 initActiveTabTracking();
 initPartnerTheme();
 initStarterBlueprints();
+initTokenRefresh();
```
Comment (johnnymetz) [inline src/background/background.ts:63]: I think `initPartnerTokenRefresh` is a better name for this function. Don't want future devs to get confused about what token we're talking about here

File: src/background/partnerIntegrations.test.ts [hunk 2]
```diff
@@ -123,3 +148,85 @@ describe("getPartnerPrincipals", () => {
...
+      {
+        id: authId,
+        serviceId: CONTROL_ROOM_OAUTH_SERVICE_ID,
+        config: {
+          controlRoomUrl: "https://controlroom.com",
+        },
+      } as unknown as RawServiceConfiguration,
+    ]);
+
+    axiosMock.onPost().reply(401);
```
Comment (johnnymetz) [inline src/background/partnerIntegrations.test.ts:224]: Just learned how to mock axios. Nice

**Review by johnnymetz — APPROVED** [review]

File: src/background/partnerIntegrations.ts [hunk 3]
```diff
@@ -131,3 +135,61 @@ export async function launchAuthIntegration({
...
+    const params = new URLSearchParams();
+    params.append("grant_type", "refresh_token");
+    params.append("client_id", context.client_id);
+    params.append("refresh_token", authData.refreshToken);
+    params.append("hosturl", config.config.controlRoomUrl);
+
+    // On 401, throw the error. In the future, we might consider clearing the partnerAuth. However, currently that
+    // would trigger a re-login, which may not be desirable at arbitrary times.
+    const { data } = await axios.post(context.tokenUrl, params, {
+      headers: { Authorization: `Basic ${btoa(context.client_id)} ` },
```
Comment (johnnymetz) [inline src/background/partnerIntegrations.ts:167]: Note to reviewer: https://developer.mozilla.org/en-US/docs/Web/API/btoa

File: src/background/partnerIntegrations.ts [hunk 4]
```diff
@@ -131,3 +135,61 @@ export async function launchAuthIntegration({
...
+    // https://axios-http.com/docs/urlencoded
+    const params = new URLSearchParams();
+    params.append("grant_type", "refresh_token");
+    params.append("client_id", context.client_id);
+    params.append("refresh_token", authData.refreshToken);
+    params.append("hosturl", config.config.controlRoomUrl);
+
+    // On 401, throw the error. In the future, we might consider clearing the partnerAuth. However, currently that
+    // would trigger a re-login, which may not be desirable at arbitrary times.
+    const { data } = await axios.post(context.tokenUrl, params, {
```
[the remark this answers is not in the document]
Comment (twschiller) [reply src/background/partnerIntegrations.ts:166]: They don't yet -- was going off of James's email

**Pull request merged**